pkg/api/v1/atlascluster_types.go

```diff
@@ -17,23 +17,231 @@ limitations under the License.
+	InstanceSizeName string `json:"instanceSizeName,omitempty"`
+
+	// Cloud service provider on which Atlas provisions the hosts.
+	// +kubebuilder:validation:Enum=AWS;GCP;AZURE;TENANT
+	ProviderName string `json:"providerName,omitempty"`
+
+	// Physical location of your MongoDB cluster.
+	// The region you choose can affect network latency for clients accessing your databases.
+	// +optional
+	RegionName string `json:"regionName,omitempty"`
+
+	// Disk IOPS setting for AWS storage.
+	// Set only if you selected AWS as your cloud service provider.
+	// +kubebuilder:validation:Enum=STANDARD;PROVISIONED
+	VolumeType string `json:"volumeType,omitempty"`
+
+	// Range of instance sizes to which your cluster can scale.
+	AutoScaling *AutoScalingSpec `json:"autoScaling,omitempty"`
+}
+
+// ReplicationSpec represents a configuration for cluster regions
+type ReplicationSpec struct {
+	// Number of shards to deploy in each specified zone.
+	// The default value is 1.
+	NumShards *int64 `json:"numShards,omitempty"`
+
+	// Name for the zone in a Global Cluster.
+	// Don't provide this value if clusterType is not GEOSHARDED.
+	// +optional
+	ZoneName string `json:"zoneName,omitempty"`
+
+	// Configuration for a region.
+	// Each regionsConfig object describes the region's priority in elections and the number and type of MongoDB nodes that Atlas deploys to the region.
+	// +optional
+	RegionsConfig map[string]RegionsConfig `json:"regionsConfig,omitempty"`
+}
+
+// RegionsConfig describes the region’s priority in elections and the number and type of MongoDB nodes Atlas deploys to the region.
+type RegionsConfig struct {
+	// The number of analytics nodes for Atlas to deploy to the region.
+	// Analytics nodes are useful for handling analytic data such as reporting queries from BI Connector for Atlas.
+	// Analytics nodes are read-only, and can never become the primary.
+	// If you do not specify this option, no analytics nodes are deployed to the region.
+	// +optional
+	AnalyticsNodes *int64 `json:"analyticsNodes,omitempty"`
+
+	// Number of electable nodes for Atlas to deploy to the region.
+	// Electable nodes can become the primary and can facilitate local reads.
+	// +optional
+	ElectableNodes *int64 `json:"electableNodes,omitempty"`
+
+	// Election priority of the region.
+	// For regions with only replicationSpecs[n].regionsConfig.<region>.readOnlyNodes, set this value to 0.
+	// +optional
+	Priority *int64 `json:"priority,omitempty"`
+
+	// Number of read-only nodes for Atlas to deploy to the region.
+	// Read-only nodes can never become the primary, but can facilitate local-reads.
+	// +optional
+	ReadOnlyNodes *int64 `json:"readOnlyNodes,omitempty"`
+}
+
+// Cluster converts the Spec to native Atlas client format.
+func (spec *AtlasClusterSpec) Cluster() *mongodbatlas.Cluster {
+	result := &mongodbatlas.Cluster{}
+	err := copier.Copy(result, spec)
```

Comment:
[q] why not json serialize -> deserialize?
I'm afraid that that library will just silently omit the non-matching fields though I believe our contract is that all `AtlasCluster` fields must match the ones from the atlas-go client?

Reply:
You're probably right, but using JSON still won't protect us from silent omissions if our tags are different for some reason - which should not happen barring a typing mistake.

We might actually have to write a unit test specifically for this - that will also give us a helpful future-proof tool to catch new fields being added to the Atlas client.

Reply:
I agree, having the unit test is the best. I think it could work in both directions: finding the fields from K8s that are not mapped to the Atlas-lib and the other direction (with some explicit exclusions for the fields we don't expose)
could you file the ticket please?

Reply:
Created [CLOUDP-79276](https://jira.mongodb.org/browse/CLOUDP-79276)